src/mqtt_client/mqtt_client.rs

```diff
@@ -112,73 +168,92 @@ where
+            let len = self.settings.get(&topic, &mut data).unwrap();
 
+            let mut prefixed_topic: String<{ 2 * MAX_TOPIC_LENGTH + 1 }> = String::new();
+            write!(&mut prefixed_topic, "{}/{}", &self.settings_prefix, &topic).unwrap();
+
+            // Note(unwrap): This should not fail because `can_publish()` was checked before
+            // attempting this publish.
+            self.mqtt
+                .client
+                .publish(
+                    &prefixed_topic,
+                    &data[..len],
+                    QoS::AtMostOnce,
+                    Retain::NotRetained,
+                    &[],
+                )
+                .unwrap();
+
+            // If we can't publish any more messages, bail out now to prevent the iterator from
+            // progressing. If we don't bail out now, we'd silently drop a setting.
             if !self.mqtt.client.can_publish(QoS::AtMostOnce) {
                 return;
             }
+        }
 
-            for topic in self
-                .settings
-                .into_iter::<{ MAX_TOPIC_LENGTH }>(iteration_state)
-                .unwrap()
-            {
-                let mut data = [0; MESSAGE_SIZE];
-
-                // Note(unwrap): We know this topic exists already because we just got it from the
-                // iterator.
-                let len = self.settings.get(&topic, &mut data).unwrap();
-
-                let mut prefixed_topic: String<{ 2 * MAX_TOPIC_LENGTH + 1 }> = String::new();
-                write!(&mut prefixed_topic, "{}/{}", &self.settings_prefix, &topic).unwrap();
-
-                // Note(unwrap): This should not fail because `can_publish()` was checked before
-                // attempting this publish.
-                self.mqtt
-                    .client
-                    .publish(
-                        &prefixed_topic,
-                        &data[..len],
-                        QoS::AtMostOnce,
-                        Retain::NotRetained,
-                        &[],
-                    )
-                    .unwrap();
-
-                // If we can't publish any more messages, bail out now to prevent the iterator from
-                // progressing. If we don't bail out now, we'd silently drop a setting.
-                if !self.mqtt.client.can_publish(QoS::AtMostOnce) {
-                    return;
-                }
-            }
+        // If we got here, we completed iterating over the topics and published them all.
+        self.state
+            .process_event(sm::Events::RepublishComplete)
+            .unwrap();
+    }
+
+    fn handle_subscription(&mut self) {
+        log::info!("MQTT connected, subscribing to settings");
 
-            iteration_exhausted = true;
+        // Note(unwrap): We construct a string with two more characters than the prefix
+        // structure, so we are guaranteed to have space for storage.
+        let mut settings_topic: String<{ MAX_TOPIC_LENGTH + 2 }> =
```

Comment:
I don't think it makes sense to fine grain this. Let's just assert that all topics are MAX_TOPIC_LENGTH.

```suggestion
        let mut settings_topic: String<{ MAX_TOPIC_LENGTH }> =
```